File: Difficulty: Medium/Subarrays with sum K/subarrays-with-sum-k.py
#User function Template for python3

class Solution:
    def countSubarrays(self, arr, k):
        # Initialize the hashmap to store prefix sums
        prefix_sum_count = {0: 1}
        prefix_sum = 0
        count = 0

        # Iterate through the array
        for num in arr:
            # Update the current prefix sum
            prefix_sum += num
            
            # Check if (prefix_sum - k) exists in the hashmap
            if prefix_sum - k in prefix_sum_count:
                count += prefix_sum_count[prefix_sum - k]
            
            # Update the hashmap with the current prefix sum
            prefix_sum_count[prefix_sum] = prefix_sum_count.get(prefix_sum, 0) + 1
        
        return count



#{ 
 # Driver Code Starts
#Initial Template for Python 3

import atexit
import io
import sys
from collections import deque

#Contributed by : Nagendra Jha

_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER


@atexit.register
def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())


if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases):
        k = int(input())
        arr = list(map(int, input().strip().split()))
        ob = Solution()
        res = ob.countSubarrays(arr, k)
        print(res)
        print("~")

# } Driver Code Ends

File: Difficulty: Medium/Subarrays with sum K/subarrays-with-sum-k.java
import java.util.*;

class Solution {
    public int cntSubarrays(int[] arr, int k) {
        Map<Integer, Integer> prefixCount = new HashMap<>();
        prefixCount.put(0, 1);  // base case: prefix sum zero before starting
        int count = 0;
        int prefixSum = 0;
        
        for (int num : arr) {
            prefixSum += num;
            
            // Check if there exists a prefix with sum (prefixSum - k)
            if (prefixCount.containsKey(prefixSum - k)) {
                count += prefixCount.get(prefixSum - k);
            }
            
            // Store/update frequency of current prefixSum
            prefixCount.put(prefixSum, prefixCount.getOrDefault(prefixSum, 0) + 1);
        }
        
        return count;
    }
}
